File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* deleteNode(TreeNode* root, int key) {
        if(!root) return root;
        else if (key < root->val) root->left = deleteNode(root->left, key);
        else if (key > root->val) root->right = deleteNode(root->right, key);
        else
        {
            if (!root->right && !root->left)
            {
                delete root;
                root = nullptr;
                return root;
            }
            else if (!root->right)
            {
                TreeNode *tmp = root;
                root = tmp->left;
                delete tmp;
                return root;
            }
            else if (!root->left)
            {
                TreeNode *tmp = root;
                root = tmp->right;
                delete tmp;
                return root;
            }
            else
            {
                TreeNode *tmp = FindMax(root->left);
                root->val = tmp->val;
                root->left =  deleteNode(root->left, tmp->val);
            }
        }
        return root;
    }
private :
    TreeNode * FindMax(TreeNode *node)
    {
        TreeNode *root = node;
        while(root->right)
            root = root->right;
        return root;
    } 
};

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} key
 * @return {TreeNode}
 */
function findMin(node)
{
    var root = node;
    while(root.left !== null) root = root.left;
    return root;
}
var deleteNode = function(root, key) {
         if(root === null) return root;
         if (root.val > key) root.left = deleteNode(root.left,key);
        else if (root.val < key) root.right =  deleteNode(root.right,key);
        else
        {
            if(root.left === null && root.right === null) root = null;
            else if (root.left === null) root = root.right;
            else if (root.right === null) root = root.left;
            else
            {
                var tmp = findMin(root.right);
                root.val = tmp.val;
                root.right = deleteNode(root.right,tmp.val);
            }
         }
   return root;
    
};